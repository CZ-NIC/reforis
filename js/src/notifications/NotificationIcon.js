/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

NotificationIcon.propTypes = {
    severity: propTypes.string.isRequired,
    className: propTypes.string,
};

export default function NotificationIcon({severity, className}) {
    let iconName = null;
    switch (severity) {
        case 'news':
            iconName = 'newspaper';
            break;
        case 'restart':
            iconName = 'power-off';
            break;
        case 'error':
            iconName = 'exclamation-circle';
            break;
        case 'update':
            iconName = 'sync';
            break;
        default:
    }

    return <i className={`fa fa-${iconName} ${className}`}/>;
}