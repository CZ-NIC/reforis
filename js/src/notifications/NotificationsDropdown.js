/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import NotificationsDropdownButton from "./NotificationsDropdownButton";
import NotificationsDropdownMenu from "./NotificationsDropdownMenu";
import NotificationsWrapper from "./NotificationsWrapper";


class NotificationsDropdown extends React.Component {
    render() {
        return <div id='notifications' className="dropdown btn-group">
            <NotificationsDropdownButton
                notificationsCount={this.props.notifications.length}
                newNotification={this.props.newNotification}

                disableNewNotification={this.props.disableNewNotification}
            />
            <NotificationsDropdownMenu
                notifications={this.props.notifications}

                dismissHandler={this.props.dismissHandler}
                dismissAllHandler={this.props.dismissAllHandler}
            />
        </div>
    }
}

export default NotificationsWrapper(NotificationsDropdown);