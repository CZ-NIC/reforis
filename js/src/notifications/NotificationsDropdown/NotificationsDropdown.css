#notifications .dropdown-menu {
    width: 20rem;
}

#notifications .dropdown-menu .btn-link {
    color: #95a5a6;
}

#notifications .dropdown-menu .btn-link:hover {
    color: #00a2e2;
    text-decoration: none;
}

#notifications .dropdown-divider-top {
    margin-bottom: 0;
}

#notifications .dropdown-divider-bottom {
    margin-top: 0;
}

#notifications .scrollable-menu {
    height: auto;
    max-height: 20rem;
    overflow-x: hidden;
}

#notifications ::-webkit-scrollbar {
    width: .5rem;
}

#notifications ::-webkit-scrollbar-thumb {
    background: #95a5a6;
    border-radius: 0.2em;
}

#notifications ::-webkit-scrollbar-thumb:hover {
    background: #4e5a5b;
}
