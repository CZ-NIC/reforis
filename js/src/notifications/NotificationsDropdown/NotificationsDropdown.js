/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';
import ReactDOM from 'react-dom';

import NotificationsDropdownButton from './NotificationsDropdownButton';
import NotificationsDropdownMenu from './NotificationsDropdownMenu';
import useNotifications, {useNewNotification} from '../hooks';


NotificationsDropdown.propTypes = {
    ws: propTypes.object.isRequired
};

export default function NotificationsDropdown({ws}) {
    const [notifications, dismiss, dismissAll] = useNotifications(ws);
    const newNotification = useNewNotification(ws);
    const container = document.getElementById('notifications_dropdown_container');
    return ReactDOM.createPortal(
        <div id='notifications' className='dropdown btn-group'>
            <NotificationsDropdownButton
                notificationsCount={notifications.length}
                newNotification={newNotification}
            />
            <NotificationsDropdownMenu
                notifications={notifications}

                dismiss={dismiss}
                dismissAll={dismissAll}
            />
        </div>,
        container,
    );
}
