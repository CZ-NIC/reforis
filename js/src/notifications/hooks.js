/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useEffect, useState } from "react";

import API_URLs from "common/API";
import { useWSForisModule, useAPIGet, useAPIPost } from "foris";

const WS_MODULE = "router_notifications";

export default function useNotifications(ws) {
    const [notifications, setNotifications] = useState([]);
    const [getState, get] = useAPIGet(API_URLs.notifications);

    useEffect(() => {
        get();
    }, [get]);
    useEffect(() => {
        if (getState.data) {
            const nonDisplayedNotifications = getState.data.notifications
                .filter((notification) => !notification.displayed);
            setNotifications(nonDisplayedNotifications);
        }
    }, [getState]);

    const [WSCreateData] = useWSForisModule(ws, WS_MODULE, "create");
    const [WSMarkAsDisplayedData] = useWSForisModule(ws, WS_MODULE, "mark_as_displayed");
    useEffect(() => {
        if (WSCreateData || WSMarkAsDisplayedData) get();
    }, [WSCreateData, WSMarkAsDisplayedData, get]);

    const [, post] = useAPIPost(API_URLs.notifications);

    function dismiss(notificationId) {
        post({ ids: [notificationId] });
        setNotifications((currentNotifications) => currentNotifications.filter(
            (notification) => notification.id !== notificationId,
        ));
    }

    function dismissAll() {
        post({ ids: notifications.map((notification) => notification.id) });
        setNotifications([]);
    }

    return [notifications, dismiss, dismissAll];
}

const NEW_NOTIFICATION_ANIMATION_DURATION = 1; // Sec.

export function useNewNotification(ws) {
    const [newNotification, setNewNotification] = useState(false);
    useEffect(() => {
        if (newNotification) {
            setTimeout(
                () => setNewNotification(false),
                NEW_NOTIFICATION_ANIMATION_DURATION * 1000,
            );
        }
    }, [newNotification]);

    const [WSCreateData] = useWSForisModule(ws, WS_MODULE, "create");

    useEffect(() => {
        if (WSCreateData) setNewNotification(true);
    }, [WSCreateData]);

    return newNotification;
}
