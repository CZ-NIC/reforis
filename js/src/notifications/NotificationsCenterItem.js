/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */
import React from 'react';
import NotificationIcon from "./NotificationIcon";
import {toLocaleDateString} from "./utils";

const BORDER_TYPES = {
    news: 'border-info',
    update: 'border-info',
    restart: 'border-danger',
    error: 'border-danger',
};

export default class NotificationsCenterItem extends React.Component {
    constructor(props) {
        super(props);
        this.myRef = React.createRef();
    }

    componentDidMount() {
        // Hack to scroll to needed item. ForisNotificationID is defined in the notification center template.
        if (window.ForisNotificationID === undefined)
            return;
        if (window.ForisNotificationID === this.props.notification.id)
            this.myRef.current.scrollIntoView({block: 'start', behavior: 'smooth'});
    }

    render() {
        const notification = this.props.notification;
        return <div ref={this.myRef} className={`card bg-light ${BORDER_TYPES[notification.severity]} sm-10`}>
            <div className="card-header">
                <NotificationIcon severity={notification.severity} className={'fa-2x'}/>
                <p className="text-muted">{toLocaleDateString(notification.created_at)}</p>
                <button
                    type="button"
                    className="close"
                    onClick={e => this.props.dismissHandler(notification.id)}
                >Ã—
                </button>
            </div>

            <div className="card-body">
                <p className="card-text">{notification.msg}</p>
            </div>
        </div>;
    }
}

