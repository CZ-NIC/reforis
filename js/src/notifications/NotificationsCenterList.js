/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, {useRef, useEffect} from 'react';
import propTypes from 'prop-types';

import {NOTIFICATION_PROP_TYPES, toLocaleDateString} from './utils';
import NotificationIcon from './NotificationIcon';

NotificationsCenterList.propTypes = {
    notifications: propTypes.arrayOf(NOTIFICATION_PROP_TYPES),
    dismiss: propTypes.func.isRequired
};

export default function NotificationsCenterList({notifications, dismiss}) {
    return notifications.map(
        notification => {
            return <NotificationsCenterItem
                key={notification.id}
                notification={notification}
                dismiss={() => dismiss(notification.id)}
            />
        }
    )
}

const BORDER_TYPES = {
    news: 'border-info',
    update: 'border-info',
    restart: 'border-danger',
    error: 'border-danger',
};

function NotificationsCenterItem({notification, dismiss}) {
    const myRef = useRef(null);

    function getIDFromURL() {
        const searchParams = new URLSearchParams(window.location.search);
        return searchParams.get('id') || null;
    }

    useEffect(() => {
        if (getIDFromURL() === notification.id)
            myRef.current.scrollIntoView({block: 'start', behavior: 'smooth'});
    }, []);


    return <div ref={myRef} className={`card bg-light ${BORDER_TYPES[notification.severity]} sm-10`}>
        <div className='card-header'>
            <NotificationIcon severity={notification.severity} className={'fa-2x'}/>
            <p className='text-muted'>{toLocaleDateString(notification.created_at)}</p>
            <button
                type='button'
                className='close'
                onClick={dismiss}
            >Ã—
            </button>
        </div>

        <div className='card-body'>
            <p className='card-text'>{notification.msg}</p>
        </div>
    </div>;
}
