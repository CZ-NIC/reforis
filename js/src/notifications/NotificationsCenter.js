/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */


import React from "react";
import NotificationsWrapper from "./NotificationsWrapper";
import NotificationsCenterItem from "./NotificationsCenterItem";

class NotificationsCenter extends React.PureComponent {
    getDismissAllButton() {
        return <button
            type="button"
            id="btn-dismiss-all"
            className="btn btn-outline-danger float-right"
            onClick={this.props.dismissAllHandler}
        >
            {_('Dismiss all')}
        </button>
    }

    render() {

        return <div id='notifications-center' className="col-sm-10">
            <h3>{_('Settings')}</h3>
            <h3>{_('Notifications')}</h3>
            {
                this.props.notifications.length !== 0 ?
                    this.getDismissAllButton() :
                    <p className="text-muted text-center">{_('No notifications')}</p>
            }

            <NotificationsCenterItemsList
                notifications={this.props.notifications}

                dismissHandler={this.props.dismissHandler}
            />
        </div>
    }
}

function NotificationsCenterItemsList({notifications, dismissHandler}) {
    return notifications.map(
        notification => {
            return <NotificationsCenterItem
                key={notification.id}
                notification={notification}
                dismissHandler={dismissHandler}
            />
        }
    )
}

export default NotificationsWrapper(NotificationsCenter);