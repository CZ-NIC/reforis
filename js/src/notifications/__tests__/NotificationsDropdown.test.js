/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, wait } from "foris/testUtils/customTestRender";

import { WebSockets } from "foris";
import mockAxios from "jest-mock-axios";
import { notificationsFixture } from "./__fixtures__/notifications";

import NotificationsDropdown from "../../main/TopBar/NotificationsDropdown/NotificationsDropdown";

describe("<NotificationsDropdown/>", () => {
    let notificationCenterContainer;

    beforeEach(async () => {
        const webSockets = new WebSockets();
        const { container, getByText } = render(<NotificationsDropdown ws={webSockets} />);
        mockAxios.mockResponse({ data: notificationsFixture });
        notificationCenterContainer = container;
        await wait(() => {
            getByText("Notification message.");
        });
    });

    it("Test with snapshot", () => {
        expect(notificationCenterContainer).toMatchSnapshot();
    });
});
