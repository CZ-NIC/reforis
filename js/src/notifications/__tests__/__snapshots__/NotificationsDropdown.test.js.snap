// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationsDropdown/> Test with snapshot 1`] = `
<div>
  <div
    class="dropdown btn-group"
    id="notifications"
  >
    <button
      class="nav-item btn btn-link"
      id="notifications-btn"
      type="button"
    >
      <span
        class="fa-stack"
      >
        <i
          class="fas fa-bell fa-stack-1x"
        />
        <div
          class=""
          id="notifications-counter"
        >
          <div
            class="number"
          >
            3
          </div>
        </div>
      </span>
    </button>
    <div
      class="dropdown-menu dropdown-menu-right dropdown-menu-lg-left"
    >
      <div
        class="dropdown-header"
        id="notifications-header"
      >
        <a
          href="/notifications"
        >
          <h5>
            Notifications
          </h5>
        </a>
        <a
          class="btn btn-link"
          href="/administration/notifications-settings"
        >
          <i
            class="fas fa-cog fa-fw"
          />
        </a>
      </div>
      <div
        class="dropdown-divider dropdown-divider-top"
      />
      <div
        class="scrollable-menu"
      >
        <div
          class="dropdown-item notification-item"
        >
          <i
            class="fa fa-newspaper fa-2x"
          />
          <div
            class="notifications-info"
          >
            <a
              href="/notifications?id=123-123"
            >
              <small
                class="text-muted"
              >
                February 1, 2000, 12:00:00 AM
              </small>
              <p>
                Notification message.
              </p>
            </a>
          </div>
          <button
            class="btn btn-link dismiss"
            type="button"
          >
            <i
              class="fas fa-times"
            />
          </button>
        </div>
        <div
          class="dropdown-divider"
        />
        <div
          class="dropdown-item notification-item"
        >
          <i
            class="fa fa-exclamation-circle fa-2x"
          />
          <div
            class="notifications-info"
          >
            <a
              href="/notifications?id=123-124"
            >
              <small
                class="text-muted"
              >
                February 1, 2000, 12:00:00 AM
              </small>
              <p>
                Second notification message.
              </p>
            </a>
          </div>
          <button
            class="btn btn-link dismiss"
            type="button"
          >
            <i
              class="fas fa-times"
            />
          </button>
        </div>
        <div
          class="dropdown-divider"
        />
        <div
          class="dropdown-item notification-item"
        >
          <i
            class="fa fa-power-off fa-2x"
          />
          <div
            class="notifications-info"
          >
            <a
              href="/notifications?id=808-909"
            >
              <small
                class="text-muted"
              >
                February 1, 2000, 12:00:00 AM
              </small>
              <p>
                Third notification message.
              </p>
            </a>
          </div>
          <button
            class="btn btn-link dismiss"
            type="button"
          >
            <i
              class="fas fa-times"
            />
          </button>
        </div>
      </div>
      <div
        class="dropdown-divider dropdown-divider-bottom"
      />
      <div
        class="dropdown-footer"
        id="notifications-footer"
      >
        <button
          class="btn btn-link"
          type="button"
        >
          Dismiss all
        </button>
      </div>
    </div>
  </div>
</div>
`;
